Internal changes to facilitate bringing library in line with other languages

mjadczak opened the pull request:
Moved the general description of changes into #263. This remains a PR only to serve as the basis upon which the full 2.0-beta goal will be achieved, and not the entire thing.

begedin commented:
@mjadczak Sorry for little feedback. I've been away for the past week or so. This looks great, but it's also huge, so It'll take a while to review thoroughly. I see the build is also failing and based on failures, I'm guessing it's still in progress? 

If you have ideas on how to maybe break it apart into smaller pieces so we can merge as we go along, that would be great. Otherwise, let us know how we can help.

JoshSmith commented:
@begedin I believe the idea as we spoke in Slack was to catch up on this PR and review it, then review PRs made against this as the base branch going forward. That way we don't have to treat the `2.0` branch as base, and can review in smaller pieces as we proceed.

Was that your understanding, too @mjadczak?

JoshSmith commented:
@mjadczak would you mind answering the question above? As more gets added to this one PR, the harder and harder it will get for us to review in any meaningful way. Could use your thoughts, too, @begedin.

mjadczak commented:
Sorry, missed this one. The issue is that I've rewritten the internal API that the surface calls use, so merging in bits isn't going to be meaningful because the result won't actually be functional until I've rewritten all the surface calls to use the new internal API for requests. I'm prioritising implementing the endpoints that are already implemented, so that at that point it makes sense to actually merge and then further endpoints just become additional features rather than outright changes and could be added in separate PRs. I'm trying to keep these changes clear and documented to allow further work by others.

I'm writing about any major decisions in the Slack channel, so that I don't go down dead ends, but in general I think the most sane way to review this is to treat it mostly as a rewrite. The major internal API changes are in `api.ex`, `entity.ex` and `request.ex`, pretty much all of the other changes at this point are going to be minor tweaks to these as needed and then pretty formulaic functions for the surface API. Here's an example of how the surface API functions now look:

```elixir
@doc """
  Create a charge.

  If your API key is in test mode, the supplied payment source (e.g., card) won't actually be
  charged, though everything else will occur as if in live mode.
  (Stripe assumes that the charge would have completed successfully).

  See the [Stripe docs](https://stripe.com/docs/api#create_charge).
  """
  @spec create(params, Stripe.options) :: {:ok, t} | {:error, Stripe.Error.t}
        when params: %{
               amount: pos_integer,
               currency: String.t,
               application_fee: non_neg_integer,
               capture: boolean,
               description: String.t,
               destination: %{
                 :account => Stripe.id | Stripe.Account.t,
                 optional(:amount) => non_neg_integer
               },
               transfer_group: String.t,
               on_behalf_of: Stripe.id | Stripe.Account.t,
               metadata: map,
               receipt_email: String.t,
               shipping: shipping,
               customer: Stripe.id | Stripe.Customer.t,
               source: Stripe.id | Stripe.Card.t | card_info,
               statement_descriptor: String.t
             }
  def create(params, opts \\ []) do
    new_request(opts)
    |> put_endpoint(@plural_endpoint)
    |> put_params(params)
    |> put_method(:post)
    |> cast_path_to_id([:destination, :account])
    |> cast_to_id([:on_behalf_of, :customer, :source])
    |> make_request()
  end
```

JoshSmith commented:
@mjadczak what I meant is that you branch off _this_ branch and we can review new pushes/PRs against this PR. That simplifies the work we have to do reviewing to make it piecemeal even if we can't get _this_ PR merged into `2.0` right away. Does that make more sense?

begedin commented:
@JoshSmith I'm not sure if I can keep proper track of changes even if we freeze this branch and start making PRs against it.

Let me try and list the changes here

# Rewrote of the core infrastructure into something more flexible

This could be a PR all by itself. Just a rewrite of how requests are built, how we use parameters, etc, all the core changes basically, including the changes to tests. Even that's huge for reviewing, but not sure it can be reduced further.

By this, I mean, basically, all the internal changes, while rewriting the existing changes, with an emphasis on keeping the old public interface, adding in deprecations for those endpoints that will have to change, could be one PR.

# Added support for missing endpoints

From there, we could work on reviewing and merging it, then start making smaller PRs for new endpoints, fixing existing endpoints, etc.

Each new endpoint, or, at least, each new module, could be a PR all by itself, easily reviewable, quick to merge.

@mjadczak I can't stress how much we appreciate the work being done here, but the reality is, it takes over an hour, even a couple of hours, to review this thoroughly, and, from the TODO list, it's not done yet, so with each new addition to this PR, the process repeats itself.

To you, since you're deep in it, it's easier to keep track of the changes, but for myself, whenever I try and review, I have to go through it all over again, thoroughly.

Would it be possible to isolate the core changes first, submit a separate PR against that, and then rebase this one once that's merged? Or at least, rebase this one against that other one?

This would be much easier if we had thorough testing before you started this amazing work, but we did not, so even keeping track of simple things such as "is this endpoint now tested", is difficult.

Just a reference, here's a list of things that changed at this point. The functions listed have been moved and rewritten, while attempting to maintain support for existing signatures for those functions, unless stated otherwise.

```
# stripe/core_resources

Balance.retrieve/1
BalanceTransaction.retrieve/2
BalanceTransaction.all/2
Charge.capture/2
Charge.create/2
Charge.retrieve/2
Charge.update/2
Charge.list/2
Customer.create/2
Customer.retrieve/2
Customer.update/3
Customer.delete/2
Customer.list/2
Event -> just moved to core resources
FileUpload - > just moved to core resources
Refund.create/2
Refund.retrieve/2
Refund.update/2
Refund.list/2
Token.create/2

# stripe/payment_methods/

Card.create/4
Card.update/5
Card/delete/4

# stripe/subscriptons

Coupon.create/2
Coupon.retrieve/2
Coupon.update/3
Coupon.delete/2
Invoice.create/2
Invoice.retrieve/2
Invoice.update/3
Invoice.pay/3
LineItem -> just moved
Plan.create/2
Plan.retrieve/2
Plan.update/3
Plan.delete/2
Subscription.create/2
Subscription.retrieve/2
Subscription.update/3
Subscription.delete/3
Subscription.list/2
Subscription.delete_discount/1
```

begedin commented on lib/stripe/entity.ex:
```diff
@@ -0,0 +1,55 @@
+defmodule Stripe.Entity do
+  @callback __from_json__(data :: map) :: map
```
This one will definitely require documentation.

We want to discourage library users from using it directly, but we also want to explain it to future contributors, so they can use it correctly when implementing new features.

begedin commented on lib/stripe/payment_methods/card.ex:
```diff
@@ -128,14 +130,21 @@ defmodule Stripe.Card do
+    do_delete(owner_type, owner_id, card_id, opts)
+  end
+
+  defp do_delete(owner_type, owner_id, card_id, opts \\ []) do
```
I'm not sure if there was a reason to split this up previously, but you might as well put the two lines in `do_delete` into `delete`, since there's no code reuse happening.

mjadczak commented:
@begedin fair points. I'll add some docs as discussed and refrain from adding any more endpoints.

JoshSmith commented:
@mjadczak sorry I haven't had a chance to look over this yet. We have a battalion ball this weekend and I'll be a bit tied up with that, but I might have some time on Sunday if I'm lucky. I would expect a (hopefully) final review no later than Tuesday from us.

tlvenn commented:
@JoshSmith any update on the review ? Would be great to review and merge if ready. Thanks in advance !

begedin commented:
@tlvenn We are working on getting this merged as quickly as we can. However, due to the volume of changes  here, we will likely have to switch from alpha to beta, and there may be incompatibilities/breaking changes.

tlvenn commented:
Thanks for the update and the revew work @begedin and @JoshSmith.

I am aware that this will likely introduce some breaking changes but that's ok with me. I have yet to adopt one stripe lib but given that we are developing a marketplace, it is essential that the lib targets the latest Stripe API. So I am kinda in the same position as @mjadczak.

And once this is merged, as you outlined @begedin, it will be easier to contribute further as well. I am also guessing that @mjadczak probably have more stuff ready but he is holding back at the moment.

mjadczak commented:
I don't plan on making any further changes to this PR, but I will base further PRs off of this one, so after this is merged into `2.0-beta` they should be much clearer and easier to review.
